### PR TITLE
Speed up Cache::retain_and_insert

### PR DESCRIPTION
**Pull request by ljedrz**
I noticed `Cache::retain_and_insert` popping up in heap profiles due to the use of `BTreeMap::split_off`, and after some investigating I realized that most of those calls can be avoided, because usually all of the contents of the cache are retained due to the cutoff time being lower (i.e. "older") than all of the timestamps in the cache.

In addition, there is an improvement to `max_cache_duplicates`, which currently costs a double database lookup and double allocation; with this trivial change, only one of each is needed to calculate it.

These changes reduced the number of allocations by ~14% in my test run.

### CODE REVIEW

File: node/narwhal/src/gateway.rs [hunk 1]
```diff
@@ -218,7 +218,7 @@ impl<N: Network> Gateway<N> {
 
     /// The maximum number of duplicates for any particular request.
     fn max_cache_duplicates(&self) -> usize {
-        self.max_committee_size() * self.max_committee_size()
+        self.max_committee_size().pow(2)
```
Comment (joske) [inline node/narwhal/src/gateway.rs:221]: wouldn't the compiler produce the same code?

Edit: checked with compiler explorer, no it doesn't, in fact it generates a lot more code in the `pow(2)` case (error checking and a call to pow - which ultimately does `imul` again). And with `opt-level = 2` it generates the exact same machine code. 

So not sure why this change makes sense?

Comment (ljedrz) [reply node/narwhal/src/gateway.rs:221]: I don't think so, as it would have no way of knowing whether the second call results in the same output

Comment (howardwu) [reply node/narwhal/src/gateway.rs:221]: can we revert since its the same?

Comment (ljedrz) [reply node/narwhal/src/gateway.rs:221]: I'll double-check it, but I'm pretty sure this is not the same, and the current form causes double calculation

Comment (howardwu) [reply node/narwhal/src/gateway.rs:221]: I see the concern about possible having a double lookup hit, I think we can keep the `pow`

**Review by howardwu — APPROVED** [review]

File: node/narwhal/src/gateway.rs (hunk 1 above)
Comment (ljedrz) [reply node/narwhal/src/gateway.rs:221]: I did a quick measurement of runtime performance of both variants under `--release`, and the current version takes ~500ns (with spikes of 1us) on average on my PC, while this change takes it down to ~250ns; both are very quick with a fresh node with an empty db, but this will become slower over time
